R/MultilabelClassifierChainsWrapper.R

```diff
@@ -0,0 +1,97 @@
+#' the feature space of each classifier is extended with true label information of all previous
+#' labels in the chain. During the prediction phase, when true labels are not available, they are
+#' replaced by predicted labels.
+#'
+#' Models can easily be accessed via \code{\link{getLearnerModel}}.
+#'
+#' @param order The chain order. E.g. for \code{m} labels, this must be a permutation of \code{1:m}. Default is \code{"random"}, i.e. \code{sample(1:m)}.
+#' @template arg_learner
+#' @template ret_learner
+#' @references
+#' Montanes, E. et al. (2013)
+#' \emph{Dependent binary relevance models for multi-label classification}
+#' Artificial Intelligence Center, University of Oviedo at Gijon, Spain.
+#' @family wrapper
+#' @family multilabel
+#' @export
+#' @examples
+#' d = getTaskData(yeast.task)
+#' # drop some labels so example runs faster
+#' d = d[, c(1:3, 15:117)]
+#' task = makeMultilabelTask(data = d, target = c("label1", "label2", "label3"))
+#' lrn = makeMultilabelClassifierChainsWrapper("classif.rpart", order = 1:3)
+#' lrn = setPredictType(lrn, "prob")
+#' # train, predict and evaluate
+#' mod = train(lrn, task)
+#' pred = predict(mod, task)
+#' p = performance(pred)
+#' performance(pred, measure = hamloss)
+#' getMultilabelBinaryPerformances(pred, measures = list(mmce, auc))
+#' # above works also with predictions from resample!
+makeMultilabelClassifierChainsWrapper = function(learner, order = "random") {
+  learner = checkLearner(learner, type = "classif")
+  if ("twoclass" %nin% learner$properties) {
+    stopf("Learner does not support binary classification")
+  }
+  id = paste("multilabel", learner$id, sep = ".")
+  packs = learner$package
+  x = makeHomogeneousEnsemble(id, learner$type, learner, packs,
+    learner.subclass = "MultilabelClassifierChainsWrapper", 
+    model.subclass = "MultilabelClassifierChainsModel")
+  x$type = "multilabel"
+  x$order = order
+  return(x)
+}
+
+#' @export
+trainLearner.MultilabelClassifierChainsWrapper = function(.learner, .task, .subset, .weights = NULL,...){
+  if (.learner$order[1] == "random") {
+    order = sample(1:length(.task$task.desc$target)) #random order
+  } else {
+    order = .learner$order
+  }
+  if (length(order) != length(getTaskTargetNames(.task))) {
+    stopf("order length does not match number of targets!")  
+  }
+  targets = getTaskTargetNames(.task)
+  .task = subsetTask(.task, subset = .subset)
+  data = getTaskData(.task)
+  models = namedList(targets[order])
+  chained_targets = targets
+  for (tn in targets[order]) {
+    data2 = dropNamed(data, setdiff(chained_targets, tn))
+    index = which(names(data2) %in% setdiff(targets, tn))
+    if (length(data2[, index]) != 0) {
+      data2[, index] = sapply(data2[, index], as.numeric)
+    }
+    ctask = makeClassifTask(id = tn, data = data2, target = tn)
+    models[[tn]] = train(.learner$next.learner, ctask, weights = .weights)
+    chained_targets = setdiff(chained_targets, tn)
```

Comment:
take this line to the beginning of the for loop, so it has not to be computed in lines 68 and 69 again